src/gas_station/gas_station_core.rs

```diff
@@ -161,10 +174,30 @@ impl GasStation {
             }
         };
         let smashed_coin_count = payment_count - updated_coins.len();
-        // Regardless of whether the transaction succeeded, we need to release the coins.
-        // Otherwise, we lose track of them. This is because `ready_for_execution` already takes
-        // the coins out of the pool and will not be covered by the auto-release mechanism.
-        self.release_gas_coins(updated_coins).await;
+        // Before returning the coins to the pool, verify if any coin exceeds
+        // NEW_COIN_BALANCE_FACTOR_THRESHOLD times the target initial balance.
+        let contains_oversized_coins = updated_coins.iter().any(|coin| {
+            coin.balance
+                > NEW_COIN_BALANCE_FACTOR_THRESHOLD * self.rescan_config.target_init_balance
+        });
+        if contains_oversized_coins {
+            warn!("Oversized coins found during transaction execution. Initiating rescan to split these coins. If this occurs frequently, consider adjusting target_init_balance or the maximum transaction budget.");
+            self.rescan_config.trigger_rescan().await;
+            self.metrics
+                .oversized_gas_coins_count
+                .with_label_values(&[&sponsor.to_string()])
+                .inc_by(updated_coins.len() as u64);
```

Comment:
Should we update by a filtered value of `updated_coins` here?

The Error match arm of `let updated_coins = match &response {` should return all (updated) coins from `payment`, regardless of size, right?

Reply:
totally right, Thanks, fixed